pytential/symbolic/pde/reduce_fmms.py

```diff
@@ -0,0 +1,320 @@
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+from sumpy.kernel import (AxisTargetDerivative, AxisSourceDerivative)
+
+from pymbolic.mapper import Mapper
+from pymbolic.primitives import Product
+from pymbolic.interop.sympy import PymbolicToSympyMapper, SympyToPymbolicMapper
+import sympy
+
+from collections import defaultdict
+
+
+__all__ = (
+    "reduce_number_of_fmms",
+    )
+
+
+def reduce_number_of_fmms(int_gs, source_dependent_variables):
+    """
+    Reduce the number of FMMs needed for a system of expressions with
+    :class:`~pytential.symbolic.primitives.IntG` objects.
+
+    This is done by converting the `IntG` expression to a matrix of polynomials
+    with d variables corresponding to d dimensions and each polynomial represents
+    a derivative operator. All the properties of derivative operator that we want
+    are reflected in the properties of the polynomial including addition,
+    multiplication and exact polynomial division.
+
+    This matrix is factored into two matrices where the left hand side matrix
+    represents a transformation at target and the right hand side matrix represents
+    a transformation at source.
+
+    If the expressions given are not linear, then the input expressions are
+    returned as is.
+
+    :arg source_dependent_variables: When reducing FMMs, consider only these
+        variables as dependent on source.
+    """
+
+    source_exprs = []
+    mapper = ConvertDensityToSourceExprCoeffMap(source_dependent_variables)
+    matrix = []
+    dim = int_gs[0].target_kernel.dim
+    axis_vars = sympy.symbols(f"_x0:{dim}")
+    to_sympy = PymbolicToSympyMapper()
+
+    if not _check_int_gs_common(int_gs):
+        return int_gs
+
+    for int_g in int_gs:
+        row = [0]*len(source_exprs)
+        for density, source_kernel in zip(int_g.densities, int_g.source_kernels):
+            try:
+                d = mapper(density)
+            except ImportError:
+                return int_gs
+            for source_expr, coeff in d.items():
+                if source_expr not in source_exprs:
+                    source_exprs.append(source_expr)
+                    row += [0]
+                poly = _convert_kernel_to_poly(source_kernel, axis_vars)
+                row[source_exprs.index(source_expr)] += poly * to_sympy(coeff)
+        matrix.append(row)
```

Comment:
Factor out the matrix finder into a function?

Reply:
Done in 7d158304bfef8dd6d6253da0c79f55c7f4706c2d